src/pymor/algorithms/line_search.py

```diff
@@ -0,0 +1,73 @@
+# This file is part of the pyMOR project (http://www.pymor.org).
+# Copyright 2013-2020 pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (http://opensource.org/licenses/BSD-2-Clause)
+
+from pymor.core.defaults import defaults
+
+
+@defaults('alpha_init', 'tau', 'beta', 'maxiter')
+def armijo(r, starting_point, correction, grad=None,
+           alpha_init=1.0, tau=0.5, beta=0.0001, maxiter=100):
+    """Armijo line search algorithm.
+
+    This method computes a step size such that the Armijo-Goldstein condition is
+    fulfilled.
+
+    Parameters
+    ----------
+    r
```

Comment:
The objective function does not have to be a residual, this could be anything. So I would not go for the name 'r' .